pkg/kubelet/kubelet.go

```diff
@@ -2845,6 +2854,110 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
 	}, 100*time.Millisecond, stopCh)
 }
 
+// handleMaxContainerCheckpoints goes through the list of checkpoints of a
+// container in a pod and deletes old checkpoints if there are more than
+// 'kl.maxContainerCheckpoints'.
+func (kl *Kubelet) handleMaxContainerCheckpoints(
+	podUID types.UID,
+	podFullName,
+	containerName string,
+	options *runtimeapi.CheckpointContainerRequest,
+) error {
+
+	if kl.maxContainerCheckpoints < 1 {
+		// This cannot really happen as the kubelet config parser will
+		// error out early if this is < 1. This additional error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxContainerCheckpoints needs to be >= 1")
+	}
+
+	globPattern := filepath.Join(
+		kl.getCheckpointsDir(),
+		fmt.Sprintf(
+			// This glob pattern tries to match RFC3339 (including leaf seconds)
+			"checkpoint-%s-%s-[0-9][0-9][0-9][0-9]-[0-2][0-9]-[0-3][0-9]T[0-2][0-9]:[0-5][0-9]:[0-6][0-9]*.tar",
+			podFullName,
+			containerName,
+		),
+	)
+
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.updateMaxContainerCheckpoints.Lock()
+	defer kl.updateMaxContainerCheckpoints.Unlock()
+
+	checkpointArchives, err := filepath.Glob(globPattern)
+
+	if err != nil {
+		return fmt.Errorf(
+			"error looking for checkpoint archives with pattern %s: %w",
+			globPattern,
+			err,
+		)
+	}
+
+	checkpointArchivesCounter := len(checkpointArchives)
+
+	if checkpointArchivesCounter <= int(kl.maxContainerCheckpoints) {
+		// Nothing to clean up, return early.
+		return nil
+	}
+
+	archivesToDelete := make(map[int64]string)
+
+	for _, c := range checkpointArchives {
+		fi, err := os.Stat(c)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to stat file %s: %w",
+				c,
+				err,
+			)
+		}
+		archivesToDelete[fi.ModTime().UnixMicro()] = c
+	}
+
+	keys := make([]int64, len(archivesToDelete))
+	i := 0
+	for key := range archivesToDelete {
+		keys[i] = key
+		i++
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
```

Comment:
Currently I am sorting it by the date returned by `stat()`. Sorting it by name would mean that it is not sorted by the actual time of the file but by the time encoded in the file name. It felt more correct to use the actual date, but I can change it to be sorted by file name, if that is preferred.

Reply:
Why can't you sort `checkpointArchives` by `date returned by stat()` ?

Reply:
> Why can't you sort `checkpointArchives` by `date returned by stat()` ?

The idea of my code was to first get a list of all checkpoint archives which belong to a certain pod/container using the glob mechanism. To not rely on the date encoded into file name I am doing a `stat()` on each file to get the time stamp on disk. Now I have a list of file names with a list of time stamps which I put in a map to have single data structure which I can use to decide which checkpoint archive to delete. I am sorting this map by key as that represents the time stamp. This way I do not have to compare timestamps I can just remove that number of checkpoint archives until it is lower than my threshold. As the map is sorted by keys (timestamps) I just need to count the number of checkpoint archives without comparing any additional time stamps.

Does that make sense?

Reply:
It makes sense, but it involves additional data structures and an extra code. Can you do the same by reverse sorting  checkpointArchives by the timestamp and then removing first `kl.maxContainerCheckpoints` archives?

Reply:
Sorry for not being clear. I meant something like this: https://go.dev/play/p/mcEIi0MYI0_V
It's a very sketchy code just to show the idea. I'd add some caching to it to avoid calling Stat multiple time for each file and may be other things, but I hope it describes the idea.

Reply:
Thanks a lot for the example. I will have a look at it.

Reply:
@adrianreber As we anyway going to call `stat` on every file, I'd suggest to use `ReadDir` instead of `Glob` and filter  files using `regexp.MatchString`. This would make the code much more clear in my opinion.

Reply:
Changed to use your example for sorting and switched to `ReadDir()` and regular expressions.